### PR TITLE
Adjust Scafi incarnation

### PR DESCRIPTION
**Pull request by metaphori**
This is to address #93 (in progress)

### CODE REVIEW

**Comment by DanySK** [comment]
Let me know as soon as this gets from "in progress" to "release candidate", I will review the code immediately

**Comment by metaphori** [comment]
@DanySK This should be fine now. Please take a look, thanks.

File: alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/implementations/nodes/ScafiNode.scala [hunk 1]
```diff
@@ -5,13 +5,12 @@
...
 
-import it.unibo.alchemist.model.implementations.nodes.AbstractNode
-import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.{Environment, Position}
 
-class ScafiNode(env: Environment[_, _]) extends AbstractNode[Any](env) {
+class ScafiNode[T,P<:Position[P]](env: Environment[T, P]) extends AbstractNode[T](env) {
 
-  override def createT = new {}
+  override def createT = throw new Exception("The molecule does not exist and cannot create empty concentration")
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/implementations/nodes/ScafiNode.scala]: Throw `IllegalStateException` or another non generic exception that you deem appropriate. Never use generic exceptions.

File: alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala [hunk 2]
```diff
@@ -61,35 +84,66 @@ sealed class ScafiIncarnation[P <: Position[P]] extends Incarnation[Any, P]{
...
+      result.setConditions(ListBuffer[Condition[T]](createCondition(rand, env, node, time, result, null)).asJava)
+
+    result
   }
 
-  override def createTimeDistribution(rand: RandomGenerator, env: Environment[Any, P], node: Node[Any], param: String): TimeDistribution[Any] = {
-    Objects.requireNonNull(param)
+  override def createTimeDistribution(rand: RandomGenerator, env: Environment[T, P], node: Node[T], param: String): TimeDistribution[T] = {
+    if (param == null) return new ExponentialTime[T](Double.PositiveInfinity, rand)
+    // Objects.requireNonNull(param, "Frequency parameter to createTimeDistribution must not be null")
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala]: Either write something to be read or remove this line

File: alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala [hunk 3]
```diff
@@ -61,35 +84,66 @@ sealed class ScafiIncarnation[P <: Position[P]] extends Incarnation[Any, P]{
...
+    val isSend = "send".equalsIgnoreCase(param)
+    val result: Reaction[T] =
+      if (isSend)
+        new ChemicalReaction[T](Objects.requireNonNull[Node[T]](node), Objects.requireNonNull[TimeDistribution[T]](time))
+      else
+        new Event[T](node, time)
+    if (param != null)
+      result.setActions(ListBuffer[Action[T]](createAction(rand, env, node, time, result, param)).asJava)
+    if (isSend)
+      result.setConditions(ListBuffer[Condition[T]](createCondition(rand, env, node, time, result, null)).asJava)
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala]: Use curly brackets. It's hard to read. Renouncing brackets is ok only in short if expressions.

File: alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala [hunk 4]
```diff
@@ -45,12 +44,36 @@ sealed class ScafiIncarnation[P <: Position[P]] extends Incarnation[Any, P]{
...
+  ) = {
+    if (!node.isInstanceOf[ScafiNode[T,P]]) {
+      throw new IllegalStateException(getClass.getSimpleName + " cannot get cloned on a node of type " + node.getClass.getSimpleName)
+    }
+    val scafiNode = node.asInstanceOf[ScafiNode[T,P]]
+    if(param=="send") {
+      val alreadyDone = ScafiIncarnationUtils.allActions[T,P,SendScafiMessage[T,P]](node, classOf[SendScafiMessage[T,P]]).map(_.program)
+        // ScafiIncarnationUtils.allScafiProgramsFor[T,P](node).filter(_.isComputationalCycleComplete)
+      val spList = ScafiIncarnationUtils.allScafiProgramsFor[T,P](node) -- alreadyDone
+
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala]: Remove empty lines within methods

Comment (DanySK) [reply alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala]: Or, if you think that some spacing is useful, write a comment instead explaining why :)

**Pull request merged**